Policy: Report debug message why inputs are non standard

ismaelsadeeq opened the pull request:
This PR is another attempt at  #13525.

Transactions that fail `PreChecks` Validation due to non-standard inputs now  returns invalid validation state`TxValidationResult::TX_INPUTS_NOT_STANDARD` along with a debug error message.

Previously, the debug error message for non-standard inputs do not specify why the inputs were considered non-standard. 
Instead, the same error string, `bad-txns-nonstandard-inputs`, used for all types of non-standard input scriptSigs.

This PR updates the `AreInputsStandard`  to include the reason why inputs are non-standard in the debug message.
This improves the `Precheck` debug message to be more descriptive.

Furthermore, I have addressed all remaining comments from #13525 in this PR.

ismaelsadeeq committed:
policy: update `AreInputsStandard` to return error string

This commit renames AreInputsStandard to ValidateInputsStandardness.

ValidateInputsStandardness now returns valid TxValidationState if all inputs
(scriptSigs) use only standard transaction forms else returns invalid
TxValidationState which states why an input is not standard.

stickies-v commented on src/policy/policy.cpp:
```diff
@@ -174,38 +175,49 @@ bool IsStandardTx(const CTransaction& tx, const std::optional<unsigned>& max_dat
  * Note that only the non-witness portion of the transaction is checked here.
  */
-bool AreInputsStandard(const CTransaction& tx, const CCoinsViewCache& mapInputs)
+std::optional<NonStandardInputsReason> AreInputsStandard(const CTransaction& tx, const CCoinsViewCache& mapInputs)
```
nit: requires `#include <optional>` (in both files)

ismaelsadeeq replied:
Fixed

stickies-v reviewed (commented):
Concept ACK

ismaelsadeeq commented on src/policy/policy.cpp:
```diff
@@ -174,38 +175,49 @@ bool IsStandardTx(const CTransaction& tx, const std::optional<unsigned>& max_dat
  * Note that only the non-witness portion of the transaction is checked here.
  */
-bool AreInputsStandard(const CTransaction& tx, const CCoinsViewCache& mapInputs)
+std::optional<NonStandardInputsReason> AreInputsStandard(const CTransaction& tx, const CCoinsViewCache& mapInputs)
```
Thank you, `HasNonStandardInput` makes more sense,  rename to `HasNonStandardInput` in 94cd47f3460

ismaelsadeeq replied:
>Alternatively, could use util::Result failure values after https://github.com/bitcoin/bitcoin/pull/25665 is merged.

will resolve this since I am now using `TxValidationState`

ismaelsadeeq commented on src/policy/policy.cpp:
```diff
@@ -174,38 +175,49 @@ bool IsStandardTx(const CTransaction& tx, const std::optional<unsigned>& max_dat
-        if (whichType == TxoutType::NONSTANDARD || whichType == TxoutType::WITNESS_UNKNOWN) {
+        if (whichType == TxoutType::NONSTANDARD) {
+            auto result = NonStandardInputsReason{"bad-txns-input-script-nonstandard", strprintf("input %u", i)};
+            return result;
```
Yes, fixed now

ismaelsadeeq commented on src/test/script_p2sh_tests.cpp:
```diff
@@ -350,10 +366,11 @@ BOOST_AUTO_TEST_CASE(AreInputsStandard)
     txTo.vin[4].scriptSig << std::vector<unsigned char>(fifteenSigops.begin(), fifteenSigops.end());
 
-    BOOST_CHECK(::AreInputsStandard(CTransaction(txTo), coins));
+    BOOST_CHECK(!::AreInputsStandard(CTransaction(txTo), coins).has_value());
```
Fixed in latest push

ismaelsadeeq commented on src/validation.cpp:
```diff
@@ -822,8 +822,9 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
-        return state.Invalid(TxValidationResult::TX_INPUTS_NOT_STANDARD, "bad-txns-nonstandard-inputs");
+    const auto inputs_standardness_result = AreInputsStandard(tx, m_view);
+    if (m_pool.m_require_standard && inputs_standardness_result.has_value()) {
+        return state.Invalid(TxValidationResult::TX_INPUTS_NOT_STANDARD, inputs_standardness_result.value().reason, inputs_standardness_result.value().debug);
```
Fixed

ismaelsadeeq commented on src/bench/ccoins_caching.cpp:
```diff
@@ -44,8 +44,8 @@ static void CCoinsCaching(benchmark::Bench& bench)
-        bool success{AreInputsStandard(tx_1, coins)};
-        assert(success);
+        auto result{AreInputsStandard(tx_1, coins)};
+        assert(!result.has_value());
```
Taken

ismaelsadeeq commented on src/policy/policy.h:
```diff
@@ -121,6 +121,11 @@ static constexpr unsigned int STANDARD_NOT_MANDATORY_VERIFY_FLAGS{STANDARD_SCRIP
 /** Used as the flags parameter to sequence and nLocktime checks in non-consensus code. */
 static constexpr unsigned int STANDARD_LOCKTIME_VERIFY_FLAGS{LOCKTIME_VERIFY_SEQUENCE};
 
+struct NonStandardInputsReason {
```
Yes, taken

ismaelsadeeq commented on src/validation.cpp:
```diff
@@ -822,8 +822,9 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
 
-    if (m_pool.m_require_standard && !AreInputsStandard(tx, m_view)) {
-        return state.Invalid(TxValidationResult::TX_INPUTS_NOT_STANDARD, "bad-txns-nonstandard-inputs");
+    const auto inputs_standardness_result = AreInputsStandard(tx, m_view);
```
Thank you fixed in 94cd47f3460

ismaelsadeeq commented:
Thank you for review @stickies-v 

I Addressed your comments and taken suggestions.
Force pushed from https://github.com/bitcoin/bitcoin/commit/6ffb9569b7f1c603ee3bfa3b56ee312d975fd8c9 to https://github.com/bitcoin/bitcoin/commit/e032462c44dba9b6961c1e18bd597d63c6afac02 , [Compare diff](https://github.com/bitcoin/bitcoin/compare/6ffb9569b7f1c603ee3bfa3b56ee312d975fd8c9..e032462c44dba9b6961c1e18bd597d63c6afac02)

luke-jr commented on src/policy/policy.cpp:
```diff
@@ -174,38 +176,44 @@ bool IsStandardTx(const CTransaction& tx, const std::optional<unsigned>& max_dat
         TxoutType whichType = Solver(prev.scriptPubKey, vSolutions);
-        if (whichType == TxoutType::NONSTANDARD || whichType == TxoutType::WITNESS_UNKNOWN) {
+        if (whichType == TxoutType::NONSTANDARD) {
+            return NonStandardInputReason{"bad-txns-input-script-nonstandard", strprintf("input %u", i)};
```
```suggestion
            return NonStandardInputReason{"bad-txns-input-script-unknown", strprintf("input %u", i)};
```

luke-jr commented on src/policy/policy.cpp:
```diff
@@ -174,38 +176,44 @@ bool IsStandardTx(const CTransaction& tx, const std::optional<unsigned>& max_dat
+                return NonStandardInputReason{"bad-txns-input-scriptsig-failure", strprintf("input %u: %s", i, ScriptErrorString(serror))};
+            }
+            if (stack.empty()) {
+                return NonStandardInputReason{"bad-txns-input-p2sh-no-redeemscript", strprintf("input %u", i)};
```
```suggestion
                return NonStandardInputReason{"bad-txns-input-scriptcheck-missing", strprintf("input %u", i)};
```

ismaelsadeeq replied:
fixed

ismaelsadeeq commented:
Thanks for reviewing @luke-jr 
Force pushed from https://github.com/bitcoin/bitcoin/commit/e032462c44dba9b6961c1e18bd597d63c6afac02 to https://github.com/bitcoin/bitcoin/commit/e032462c44dba9b6961c1e18bd597d63c6afac02.

[Compare diff](https://github.com/bitcoin/bitcoin/compare/e032462c44dba9b6961c1e18bd597d63c6afac02..814571bf3ebb4e2640bbb6401c5822c038bb5287)

The changes in latest push are:
- Rebased on master to fix CI
- Updated the non-standardness reason to address @luke-jr comments
- `bad-txns-input-script-nonstandard` --> `bad-txns-input-script-unknown`
- `bad-txns-input-scriptsig-failure` --> `bad-txns-input-p2sh-scriptsig-malformed`
- `bad-txns-input-p2sh-no-redeemscript` --> `bad-txns-input-scriptcheck-missing`
- `bad-txns-input-scriptcheck-sigops` --> `bad-txns-input-p2sh-redeemscript-sigops`

glozow commented on src/policy/policy.h:
```diff
@@ -121,6 +122,11 @@ static constexpr unsigned int STANDARD_NOT_MANDATORY_VERIFY_FLAGS{STANDARD_SCRIP
 /** Used as the flags parameter to sequence and nLocktime checks in non-consensus code. */
 static constexpr unsigned int STANDARD_LOCKTIME_VERIFY_FLAGS{LOCKTIME_VERIFY_SEQUENCE};
 
+struct NonStandardInputReason {
+    std::string reason;
+    std::string debug;
+};
```
Maybe just use `TxValidationState`?

ismaelsadeeq replied:
Thanks updated

luke-jr commented on src/validation.cpp:
```diff
@@ -848,8 +848,11 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
-    if (m_pool.m_opts.require_standard && !AreInputsStandard(tx, m_view)) {
-        return state.Invalid(TxValidationResult::TX_INPUTS_NOT_STANDARD, "bad-txns-nonstandard-inputs");
+    if (m_pool.m_opts.require_standard) {
+        state = HasNonStandardInput(tx, m_view);
```
Seems like it'd be nicer to just pass `state` into the function rather than creating a new one and copying it.

ismaelsadeeq replied:
> Seems like it'd be nicer to just pass state into the function 

I am taking this review comment as a non-blocking (nit) suggestion based on the tone.

I will update this when there are blocking comments that indicate a need to retouch the diffs.

Also, I would like to know whether this PR is conceptually acknowledged, @luke-jr.
Thanks for your review.

ismaelsadeeq replied:
In this case, a new state is indeed created in `HasNonStandardInput`, but either RVO or an implicit move is performed not a copy?

I think, in general, returning a value is preferred over using out parameters?
same was also recommended here https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#coding-style-c-functions-and-methods

achow101 commented:
cc @tdb3

tdb3 commented on src/test/script_p2sh_tests.cpp:
```diff
@@ -333,7 +333,23 @@ BOOST_AUTO_TEST_CASE(HasNonStandardInput)
+
+    // vout [8] is non-standard because it contains OP_RETURN in its scriptSig.
+    static const unsigned char op_return[] = {OP_RETURN};
+    txFrom.vout[8].scriptPubKey = GetScriptForDestination(ScriptHash(CScript(op_return, op_return + sizeof(op_return))));
```
Instantiating the CScript here could allow re-use later (in `txToNonStd4.vin[0].scriptSig = CScript(op_return, op_return + sizeof(op_return));`)

tdb3 reviewed (commented):
I'm a fan of the more descriptive errors.

However, would some of these more descriptive errors flow down to RPC responses and cause issues for users? (e.g. `testmempoolaccept`).

https://github.com/bitcoin/bitcoin/pull/30212#issuecomment-2347371722
> Does bitcoind consider this to be a breaking change? The new error message may be "more accurate", but for clients that were matching on this error to figure out why a transaction was rejected, this breaks old behavior.

ismaelsadeeq commented on src/policy/policy.h:
```diff
@@ -139,11 +140,12 @@ static constexpr decltype(CTransaction::version) TX_MAX_STANDARD_VERSION{3};
+ * Check for standard transaction types
+ * @param[in] mapInputs       Map of previous transactions that have outputs we're spending
+ * @returns valid TxValidationState if all inputs (scriptSigs) use only standard transaction forms else returns
+ * invalid TxValidationState which states why an input is not standard.
```
Done thanks

ismaelsadeeq commented:
> > Does bitcoind consider this to be a breaking change? The new error message may be "more accurate", but for clients that were matching on this error to figure out why a transaction was rejected, this breaks old behavior.

I've added a release notes for this

tdb3 commented:
> > > Does bitcoind consider this to be a breaking change? The new error message may be "more accurate", but for clients that were matching on this error to figure out why a transaction was rejected, this breaks old behavior.
> 
> I've added a release notes for this

The release note helps a lot. To err on the side of caution, it seems appropriate to include a `deprecatedrpc=` option, to enable a period of deprecation for users.

ismaelsadeeq commented:
> The release note helps a lot. To err on the side of caution, it seems appropriate to include a `-deprecatedrpc=` option, to enable a period of deprecation for users.

Thanks for your review @tdb3  will address this comment soon.

ismaelsadeeq commented:
> The release note helps a lot. To err on the side of caution, it seems appropriate to include a `-deprecatedrpc=` option, to enable a period of deprecation for users.

I attempted this, but it's  a bit non trivial with some if-else branching. I also had to add a new `TransactionError` enum type to handle `sendrawtransaction` case.
<details>
<summary> see **untested** diff </summary>

```diff
diff --git a/src/common/messages.cpp b/src/common/messages.cpp
index 5fe3e9e4d86..3142ca07b4c 100644
--- a/src/common/messages.cpp
+++ b/src/common/messages.cpp
@@ -132,6 +132,8 @@ bilingual_str TransactionErrorString(const TransactionError err)
             return Untranslated("Transaction outputs already in utxo set");
         case TransactionError::MEMPOOL_REJECTED:
             return Untranslated("Transaction rejected by mempool");
+        case TransactionError::MEMPOOL_REJECTED_NON_STANDARD_INPUT:
+            return Untranslated("Transaction rejected by mempool");
         case TransactionError::MEMPOOL_ERROR:
             return Untranslated("Mempool internal error");
         case TransactionError::MAX_FEE_EXCEEDED:
diff --git a/src/node/transaction.cpp b/src/node/transaction.cpp
index 0f45da45dbd..c35bee4ee61 100644
--- a/src/node/transaction.cpp
+++ b/src/node/transaction.cpp
@@ -25,6 +25,9 @@ static TransactionError HandleATMPError(const TxValidationState& state, std::str
         if (state.GetResult() == TxValidationResult::TX_MISSING_INPUTS) {
             return TransactionError::MISSING_INPUTS;
         }
+        if (state.GetResult() == TxValidationResult::TX_INPUTS_NOT_STANDARD) {
+            return TransactionError::MEMPOOL_REJECTED_NON_STANDARD_INPUT;
+        }
         return TransactionError::MEMPOOL_REJECTED;
     } else {
         return TransactionError::MEMPOOL_ERROR;
diff --git a/src/node/types.h b/src/node/types.h
index 1302f1b127f..5cdb696204f 100644
--- a/src/node/types.h
+++ b/src/node/types.h
@@ -21,6 +21,7 @@ enum class TransactionError {
     MISSING_INPUTS,
     ALREADY_IN_UTXO_SET,
     MEMPOOL_REJECTED,
+    MEMPOOL_REJECTED_NON_STANDARD_INPUT,
     MEMPOOL_ERROR,
     MAX_FEE_EXCEEDED,
     MAX_BURN_EXCEEDED,
diff --git a/src/rpc/mempool.cpp b/src/rpc/mempool.cpp
index d61898260bf..a9f434fa605 100644
--- a/src/rpc/mempool.cpp
+++ b/src/rpc/mempool.cpp
@@ -95,6 +95,9 @@ static RPCHelpMan sendrawtransaction()
             NodeContext& node = EnsureAnyNodeContext(request.context);
             const TransactionError err = BroadcastTransaction(node, tx, err_string, max_raw_tx_fee, /*relay=*/true, /*wait_callback=*/true);
             if (TransactionError::OK != err) {
+                if (IsDeprecatedRPCEnabled("nonstandard-inputs-err") && err == TransactionError::MEMPOOL_REJECTED_NON_STANDARD_INPUT) {
+                    throw JSONRPCTransactionError(err, "bad-txns-nonstandard-inputs");
+                }
                 throw JSONRPCTransactionError(err, err_string);
             }
 
@@ -241,7 +244,11 @@ static RPCHelpMan testmempoolaccept()
                     if (state.GetResult() == TxValidationResult::TX_MISSING_INPUTS) {
                         result_inner.pushKV("reject-reason", "missing-inputs");
                     } else {
-                        result_inner.pushKV("reject-reason", state.GetRejectReason());
+                        if (IsDeprecatedRPCEnabled("nonstandard-inputs-err") && state.GetResult() == TxValidationResult::TX_INPUTS_NOT_STANDA
RD) {
+                            result_inner.pushKV("reject-reason", "bad-txns-nonstandard-inputs");
+                        } else {
+                            result_inner.pushKV("reject-reason", state.GetRejectReason());
+                        }
                     }
                 }
                 rpc_result.push_back(std::move(result_inner));
@@ -961,7 +968,6 @@ static RPCHelpMan submitpackage()
             }
 
             UniValue rpc_result{UniValue::VOBJ};
-            rpc_result.pushKV("package_msg", package_msg);
             UniValue tx_result_map{UniValue::VOBJ};
             std::set<uint256> replaced_txids;
             for (const auto& tx : txns) {
@@ -979,7 +985,11 @@ static RPCHelpMan submitpackage()
                     result_inner.pushKV("other-wtxid", it->second.m_other_wtxid.value().GetHex());
                     break;
                 case MempoolAcceptResult::ResultType::INVALID:
-                    result_inner.pushKV("error", it->second.m_state.ToString());
+                    if (IsDeprecatedRPCEnabled("nonstandard-inputs-err") && it->second.m_state.GetResult() == TxValidationResult::TX_INPUTS_N
OT_STANDARD){
+                        result_inner.pushKV("error", "bad-txns-nonstandard-inputs");
+                    } else {
+                        result_inner.pushKV("error", it->second.m_state.ToString());
+                    }
                     break;
                 case MempoolAcceptResult::ResultType::VALID:
                 case MempoolAcceptResult::ResultType::MEMPOOL_ENTRY:
diff --git a/src/rpc/util.cpp b/src/rpc/util.cpp
index 678bac7a185..3653421dcae 100644
--- a/src/rpc/util.cpp
+++ b/src/rpc/util.cpp
@@ -400,6 +400,8 @@ RPCErrorCode RPCErrorFromTransactionError(TransactionError terr)
     switch (terr) {
         case TransactionError::MEMPOOL_REJECTED:
             return RPC_TRANSACTION_REJECTED;
+        case TransactionError::MEMPOOL_REJECTED_NON_STANDARD_INPUT:
+            return RPC_TRANSACTION_REJECTED;
         case TransactionError::ALREADY_IN_UTXO_SET:
             return RPC_VERIFY_ALREADY_IN_UTXO_SET;
         default: break;
diff --git a/src/test/fuzz/kitchen_sink.cpp b/src/test/fuzz/kitchen_sink.cpp
index 62b49106cdd..4df4a2c4372 100644
--- a/src/test/fuzz/kitchen_sink.cpp
+++ b/src/test/fuzz/kitchen_sink.cpp
@@ -25,6 +25,7 @@ constexpr TransactionError ALL_TRANSACTION_ERROR[] = {
     TransactionError::MISSING_INPUTS,
     TransactionError::ALREADY_IN_UTXO_SET,
     TransactionError::MEMPOOL_REJECTED,
+    TransactionError::MEMPOOL_REJECTED_NON_STANDARD_INPUT,
     TransactionError::MEMPOOL_ERROR,
     TransactionError::MAX_FEE_EXCEEDED,
 };
```
</details>

However, do you think it would be better to make the prefix of the message `bad-txns-nonstandard-inputs` and append the additional information after it? 
This approach ensures that clients relying solely on pattern matching remain unaffected, as the error code remains unchanged.

The only clients that affected after the above suggestion are those performing an exact match. 

Let me know what you think about the two approaches.

ismaelsadeeq commented:
Rebased and addressed comment from @DrahtBot see [diff](https://github.com/bitcoin/bitcoin/compare/b487edcffab9ace19fd927998d82e945bd39b0a4..3cde7f185255b1bc07a90e95708665e0b6d5a46d)

ismaelsadeeq commented:
Rebased after the merge of #32521 to pick up recent changes to `AreInputsStandard`.

A new verbose message will be returned when the BIP54 legacy sigops limit is violated:

* `bad-txns-non-witness-sigops-exceeds-bip54-limit`: This occurs when the total number of non-witness sigops across the entire transaction exceeds `MAX_TX_LEGACY_SIGOPS`. Additionally, a debug message is also returned with the actual number of sigops and the limit, for example: `sigops 2501 > limit 2500`.


[3cde7f185255b1bc07a90e95708665e0b6d5a46d..85c6c97b609bcbcd2b2f13638d858d5a31a903](https://github.com/bitcoin/bitcoin/compare/3cde7f185255b1bc07a90e95708665e0b6d5a46d..85c6c97b609bcbcd2b2f13638d858d5a31a90379)

sedited reviewed (commented):
Concept ACK

There are also some drahtbot llm linter suggestions.

ismaelsadeeq committed:
test: ensure `ValidateInputsStandardness` optionally returns debug string

- Test the various input scriptSig non standardness, and ensure ValidateInputsStandardness
  Returns the reason why the input is nonstandard.

Co-authored-by: Ben Woosley <ben.woosley@gmail.com>

ismaelsadeeq committed:
doc: add release notes

sipa reviewed (commented):
utACK 9eea72d3f3647197c24329b412c7fc71895e3ea2

sedited reviewed (approved):
ACK 9eea72d3f3647197c24329b412c7fc71895e3ea2

ismaelsadeeq commented on src/policy/policy.cpp:
```diff
@@ -210,42 +211,56 @@ static bool CheckSigopsBIP54(const CTransaction& tx, const CCoinsViewCache& inpu
 
-    if (!CheckSigopsBIP54(tx, mapInputs)) {
-        return false;
+    auto sigops_check_result = CheckSigopsBIP54(tx, mapInputs);
```
Yes, fixed.

sedited reviewed (approved):
Re-ACK b4db7f3dbaffab93cbb0e43da7b482fc194a4a74

instagibbs reviewed (commented):
concept ACK

ismaelsadeeq commented:
Forced pushed from b4db7f3dbaffab93cbb0e43da7b482fc194a4a74 to d8f4e7caf0d25aaf4b3978d6ba698659b46473b4  [compare diff](https://github.com/bitcoin/bitcoin/compare/b4db7f3dbaffab93cbb0e43da7b482fc194a4a74..d8f4e7caf0d25aaf4b3978d6ba698659b46473b4)

- Addressed review comments from @instagibbs thanks https://github.com/bitcoin/bitcoin/pull/29060#discussion_r2872800651  Taken with a modification, https://github.com/bitcoin/bitcoin/pull/29060#discussion_r2872832879 Taken, https://github.com/bitcoin/bitcoin/pull/29060#discussion_r2872842277 Clarified.

sedited reviewed (approved):
Re-ACK d8f4e7caf0d25aaf4b3978d6ba698659b46473b4

darosior commented on src/test/transaction_tests.cpp:
```diff
@@ -1083,8 +1091,7 @@ BOOST_AUTO_TEST_CASE(max_standard_legacy_sigops)
     // The transaction now contains exactly 2500 sigops, the check should pass.
-    BOOST_CHECK_EQUAL(p2sh_inputs_count * MAX_P2SH_SIGOPS + p2pk_inputs_count * 1, MAX_TX_LEGACY_SIGOPS);
-    BOOST_CHECK(::AreInputsStandard(CTransaction(tx_max_sigops), coins));
+    BOOST_CHECK(::ValidateInputsStandardness(CTransaction(tx_max_sigops), coins).IsValid());
```
Why did this remove the assertion on the number of sigops? This is useful documentation, plus useful in making sure the comment above the check does not become stale or incorrect.

ismaelsadeeq replied:
Thanks for spotting this; fixed here https://github.com/bitcoin/bitcoin/pull/35328

darosior commented on src/test/script_p2sh_tests.cpp:
```diff
@@ -391,8 +413,67 @@ BOOST_AUTO_TEST_CASE(AreInputsStandard)
+    BOOST_CHECK(txToNonStd2_no_scriptSig_res.IsInvalid());
+    BOOST_CHECK_EQUAL(txToNonStd2_no_scriptSig_res.GetRejectReason(), "bad-txns-nonstandard-inputs");
+    BOOST_CHECK_EQUAL(txToNonStd2_no_scriptSig_res.GetDebugMessage(), "input 0 P2SH redeemscript missing");
+    BOOST_CHECK_EQUAL(GetP2SHSigOpCount(CTransaction(txToNonStd2), coins), 20U);
```
See https://github.com/bitcoin/bitcoin/pull/35863
